File: Leetcode/_121_Best_Time_to_Buy_and_Sell_Stock.java
// link - https://leetcode.com/problems/best-time-to-buy-and-sell-stock/

class _121_Best_Time_to_Buy_and_Sell_Stock {
    // Prefix/Postfix sum solution - Time = O(3n) = O(n) - Space = O(2n) = O(n)
    public int maxProfit(int[] prices) {
        int n = prices.length;
        int[] prefixArr = new int[n];
        int[] postfixArr = new int[n];

        prefixArr[0] = prices[0];
        for (int i = 1; i < n; i++) {
            prefixArr[i] = Math.min(prefixArr[i - 1], prices[i]);
        }

        postfixArr[n - 1] = prices[n - 1];
        for (int i = n - 2; i >= 0; i--) {
            postfixArr[i] = Math.max(postfixArr[i + 1], prices[i]);
        }

        int maxDiff = 0;
        for (int i = 0; i < n; i++) {
            maxDiff = Math.max(maxDiff, postfixArr[i] - prefixArr[i]);
        }

        return maxDiff;
    }
}

File: Leetcode/python/_121_Best_Time_to_Buy_and_Sell_Stock.py
from typing import List


class Solution:
    def maxProfit(self, prices: List[int]) -> int:
        max_profit = 0

        i, j = 0, 1
        while j < len(prices):
            if prices[j] < prices[i]:
                i = j
                j += 1
                continue

            max_profit = max(prices[j] - prices[i], max_profit)
            j += 1

        return max_profit

    # Greedy version - Time = O(n) - Space = O(1)
    # What we do is - initially we consider prices[0] as out buy price, since thats the only price we're guarenteed to have
    # And then we iterate over each price and check for
    #   - if the current profit difference is greater than max_profit then update max_profit with that value
    #   - if current price is smaller than our buy price then change buy price to that, since "buy low(est) sell high(est) = $$$"
    # We check for these 2 cases exclusively because if we find a lower value for buy_price, then we'll consider the price difference
    # for the element after the updated buy_price element
    def maxProfit(self, prices: List[int]) -> int:
        buy_price = prices[0]
        max_profit = 0

        for price in prices:
            if price < buy_price:
                buy_price = price
            elif price - buy_price > max_profit:
                max_profit = price - buy_price

        return max_profit
